group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java

```diff
@@ -791,6 +819,60 @@ private void validateMemberEpoch(
         }
     }
 
+    /**
+     * Computes the subscription type based on the provided information.
+     *
+     * @param subscribedRegularExpressions  The subscribed regular expression count.
+     * @param subscribedTopicNames          The subscribed topic name count.
+     * @param numberOfMembers               The number of members in the group.
+     *
+     * @return The subscription type.
+     */
+    public static SubscriptionType subscriptionType(
+        Map<String, Integer> subscribedRegularExpressions,
+        Map<String, Integer> subscribedTopicNames,
+        int numberOfMembers
+    ) {
+        if (subscribedRegularExpressions.isEmpty()) {
+            // If the members do not use regular expressions, the subscription is
+            // considered as homogeneous if all the members are subscribed to the
+            // same topics. Otherwise, it is considered as heterogeneous.
+            for (int subscriberCount : subscribedTopicNames.values()) {
+                if (subscriberCount != numberOfMembers) {
+                    return HETEROGENEOUS;
+                }
+            }
+            return HOMOGENEOUS;
+        } else {
+            int count = subscribedRegularExpressions.values().iterator().next();
+            if (count == numberOfMembers) {
+                // If all the members are subscribed to a single regular expressions
+                // and none of them are subscribed to topic names, the subscription
+                // is considered as homogeneous. If some members are subscribed to
+                // topic names too, the subscription is considered as heterogeneous.
+                for (int subscriberCount : subscribedTopicNames.values()) {
+                    if (subscriberCount != 1) {
+                        return HETEROGENEOUS;
+                    }
+                }
+                return HOMOGENEOUS;
```

Comment:
I'm a little confused about this. Won't we return `HOMOGENEOUS` when all members are subscribed to the same regex, but one member is additionally subscribed to a topic outside the scope of the regex?
eg. all members subscribe to regex "foo*" and one member also subscribes to topic "bar".

Reply:
good catch! i did not consider this case. let me see how i can fix it.

Reply:
@squah-confluent I updated the PR to address it. I opted for tracking the number of subscribed names and of subscribed regexes separately in order to avoid it. Let me know what you think.

Reply:
looks great! the fix makes sense